Revise down the nondeterminism tutorial (#1353)
model-checking/kani: docs/src/tutorial-nondeterministic-variables.md

Here is a simple example (available [here](https://github.com/model-checking/kani/blob/main/docs/src/tutorial/arbitrary-variables/src/inventory.rs)):

```rust
{{#include tutorial/arbitrary-variables/src/inventory.rs:inventory_lib}}
```

Let's write a fairly simple proof harness, one that just ensures we successfully `get` the value we inserted with `update`:

```rust
{{#include tutorial/arbitrary-variables/src/inventory.rs:safe_update}}
```

We use `kani::any()` twice here:

1. `id` has type `ProductId` which was actually just a `u32`, and so any value is fine.
2. `quantity`, however, has type `NonZeroU32`.
In Rust, it would be undefined behavior to have a value of `0` for this type.

We included an extra assertion that the value returned by `kani::any()` here was actually non-zero.
If we run this, you'll notice that verification succeeds.

```bash
cargo kani --harness safe_update
```

`kani::any()` is safe Rust, and so Kani only implements it for types where type invariants are enforced.
For `NonZeroU32`, this means we never return a `0` value.
The assertion we wrote in this harness was just an extra check we added to demonstrate this fact.

> ## Other kinds of nondeterministic variables

Reviewer: What are you trying to say in this section? I'm tempted to just remove this for now.

Author: I think this structure is good bones, but I see what you're saying about it seeming pointless right now.

As soon as we decide things like `any_vec` are stable and we commit to their API, this is where I'd put it. So this section was kinda meant to be about that until I went and audited the docs and felt that maybe we wanted to revise that API first...

Reviewer 2: I agree this section is a bit confusing. Not sure what it's getting at.

There's nothing special about `kani::any()` except that Kani ships with implementations for a few types where we can guarantee safety.

Note that there is no implementation for `Vec<T>`, however.
The trouble with a nondeterministic vector is that you usually need to _bound_ the size of the vector, for the reasons we investigated in the [last chapter](./tutorial-loop-unwinding.md).
There are no arguments provided to `kani::any()` to indicate an upper bound.

You can use `kani::any()` for `[T; N]` (if implemented for `T`).
But this the type of an exactly-sized array.
Again, `kani::any()` does not have an implementation for an unsized slice (like `[T]`).

This does not mean you cannot have a nondeterministic vector.
It just means you have to construct one.
Our example proof harness above constructs a nondeterministic `Inventory` of size `1`, simply by starting with the empty `Inventory` and inserting a nondeterministic entry.

## Custom nondeterministic types

> When you need nondeterministic variables of types that `kani::any()` cannot construct, our current recommendation is to simply write a function.

Reviewer: Why not mention the Arbitrary type?

Author: That was my original plan, and then Nathan tried it and immediately ran into the lack of `Option<MyType>`.

So I've noted that in the associate issue and was just going to leave it as-is for now.

Reviewer 2: Why does the lack of `Option<MyType>` affect the `Arbitrary` type story though?

Reviewer: https://github.com/model-checking/kani/pull/1401

To generate a nondeterministic struct, you would just generate nondeterministic values for each of its fields.
For complex data structures, you can start with an empty one and add a (bounded) nondeterministic number of entries.
For an enum, you can make use of a simple trick:

```rust
{{#include tutorial/arbitrary-variables/src/rating.rs:rating_invariant}}
```

All we're doing here is making use of a nondeterministic integer to decide which variant of `Rating` to return.

> **NOTE**: If we thought of this code as generating a random value, this function looks sub-optimal.
> We'd overwhelming generate a `Three` because it's matching "all other integers besides 1 and 2."
> But Kani just see 3 meaningful possibilities, each of which is not treated any differently from each other.
> The "proportion" of integers does not matter.

> ## Exercise

Reviewer: I like this. Should we provide an answer for this? Maybe refer to a file that they can open at the end.

# Nondeterministic variables

Kani is able to reason about programs and their execution paths by allowing users to assign nondeterministic (also called symbolic) values to certain variables using `kani::any()`.
Kani is a "bit-precise" model checker, which means that Kani considers all the possible bit-value combinations that could be assigned to the variable's memory contents.

As a Rust developer, this sounds a lot like the `mem::transmute` operation, which is highly `unsafe`.
And that's correct.

> But `kani::any()` is only implemented for a few types: those where we are able to safely enforce the type's invariants.

Reviewer 2: I think this will beg the question: "What should I do if I need `kani::any()` for one of the unimplemented types?" So perhaps add a reference to the section that discusses that, e.g:

(See Section X for a discussion on how to assign nondeterministic values to types that `kani::any()` is not implemented for)

> In other words, `kani::any()` should not produce values that are invalid for the type.

Reviewer 2: Join this sentence with the previous paragraph?

You can use `kani::any()` for `[T; N]` (if implemented for `T`).
But this the type of an exactly-sized array.
Again, `kani::any()` does not have an implementation for an unsized slice (like `[T]`).

This does not mean you cannot have a nondeterministic vector.
It just means you have to construct one.
Our example proof harness above constructs a nondeterministic `Inventory` of size `1`, simply by starting with the empty `Inventory` and inserting a nondeterministic entry.

## Custom nondeterministic types

When you need nondeterministic variables of types that `kani::any()` cannot construct, our current recommendation is to simply write a function.

To generate a nondeterministic struct, you would just generate nondeterministic values for each of its fields.
For complex data structures, you can start with an empty one and add a (bounded) nondeterministic number of entries.
For an enum, you can make use of a simple trick:

```rust
{{#include tutorial/arbitrary-variables/src/rating.rs:rating_invariant}}
```

All we're doing here is making use of a nondeterministic integer to decide which variant of `Rating` to return.

> > **NOTE**: If we thought of this code as generating a random value, this function looks sub-optimal.

Reviewer 2: ```suggestion
> **NOTE**: If we thought of this code as generating a random value, this function looks heavily biased.
```

> > We'd overwhelming generate a `Three` because it's matching "all other integers besides 1 and 2."

Reviewer 2: ```suggestion
> We'd overwhelmingly generate a `Three` because it's matching "all other integers besides 1 and 2."
```

> But Kani just see 3 meaningful possibilities, each of which is not treated any differently from each other.

> > The "proportion" of integers does not matter.

Reviewer 2: Nice note!

## Exercise

> Try writing a function to generate a nondeterministic inventory (from the first example:)

Reviewer 2: ```suggestion
Try writing a function to generate a nondeterministic inventory of a bounded size (from the first example:)
```
?

While `kani::any()` is the only method Kani provides to inject non-determinism into a proof harness, Kani only ships with implementations for a few types where we can guarantee safety.
When you need nondeterministic variables of types that `kani::any()` cannot construct, you have two options:

1. Implement the `kani::Arbitrary` trait for your type, so you can use `kani::any()`.
2. Just write a function.

The advantage of the first approach is that it's simple and conventional.
It also means that in addition to being able to use `kani::any()` with your type, you can also use it with `Option<MyType>` (for example).

The advantage of the second approach is that you're able to pass in parameters, like bounds on the size of the data structure.
(Which we'll discuss more in the next section.)
This approach is also necessary when you are unable to implement a trait (like `Arbitrary`) on a type you're importing from another crate.

Either way, inside this function you would simply return an arbitrary value by generating arbitrary values for its components.
To generate a nondeterministic struct, you would just generate nondeterministic values for each of its fields.

> For complex data structures, you can start with an empty one and add a (bounded) nondeterministic number of entries.

Reviewer: Do you mean for containers?

# Nondeterministic variables

Kani is able to reason about programs and their execution paths by allowing users to assign nondeterministic (also called symbolic) values to certain variables using `kani::any()`.

> Kani is a "bit-precise" model checker, which means that Kani considers all the possible bit-value combinations that could be assigned to the variable's memory contents.

Reviewer: what about saying 
> ... possible bit-value combinations that would be valid if assigned to the variable's memory contents. In other words, `kani::any()` should not produce values that are invalid for the type.

```

We use `kani::any()` twice here:

1. `id` has type `ProductId` which was actually just a `u32`, and so any value is fine.
2. `quantity`, however, has type `NonZeroU32`.
In Rust, it would be undefined behavior to have a value of `0` for this type.

We included an extra assertion that the value returned by `kani::any()` here was actually non-zero.
If we run this, you'll notice that verification succeeds.

```bash
cargo kani --harness safe_update
```

`kani::any()` is safe Rust, and so Kani only implements it for types where type invariants are enforced.
For `NonZeroU32`, this means we never return a `0` value.
The assertion we wrote in this harness was just an extra check we added to demonstrate this fact.

## Bounding nondeterministic variables

There is no implementation of `kani::any()` for more complex data structures like `Vec`, however.
The trouble with a nondeterministic vector is that you usually need to _bound_ the size of the vector, for the reasons we investigated in the [last chapter](./tutorial-loop-unwinding.md).
The `kani::any()` function does not have any arguments, and so cannot be given an upper bound.

> You can use `kani::any()` for `[T; N]` (if implemented for `T`).

Reviewer: This line doesn't make much sense here.

What if you move talk about custom non-deterministic types first then you introduce this section. I think it is easier to first present the solution and discuss its limitation after.

But this the type of an exactly-sized array.
Again, `kani::any()` does not have an implementation for an unsized slice (like `[T]`).

This does not mean you cannot have a nondeterministic vector.
It just means you have to construct one.
Our example proof harness above constructs a nondeterministic `Inventory` of size `1`, simply by starting with the empty `Inventory` and inserting a nondeterministic entry.

## Custom nondeterministic types

> There's nothing special about `kani::any()` except that Kani ships with implementations for a few types where we can guarantee safety.

Reviewer: maybe rephrase this to something like:
> `kani::any()` is the only method that Kani provides to inject non-determinism in a harness, but Kani ships with implementations for just a few types where it can guarantee safety.

When you need nondeterministic variables of types that `kani::any()` cannot construct, you have two options:

1. Implement `kani::Arbitrary` for your type, so you can use `kani::any()`.
2. Just write a function.

The advantage of the first approach is that it's simple and conventional.
It also means that in addition to being able to use `kani::any()` with your type, you can also use it with `Option<MyType>` (for example).

> The advantage of the second approach is that you're able to pass in parameters, like bounds on the size of the data structure.

Reviewer: Here you mention the parameter, so you could just replace the previous section by giving a bit more details on the second approach. Maybe add an example.

```rust
{{#include tutorial/arbitrary-variables/src/inventory.rs:inventory_lib}}
```

Let's write a fairly simple proof harness, one that just ensures we successfully `get` the value we inserted with `update`:

```rust
{{#include tutorial/arbitrary-variables/src/inventory.rs:safe_update}}
```

We use `kani::any()` twice here:

1. `id` has type `ProductId` which was actually just a `u32`, and so any value is fine.
2. `quantity`, however, has type `NonZeroU32`.
In Rust, it would be undefined behavior to have a value of `0` for this type.

We included an extra assertion that the value returned by `kani::any()` here was actually non-zero.
If we run this, you'll notice that verfication succeeds.

```bash
cargo kani --harness safe_update
```

`kani::any()` is safe Rust, and so Kani only implements it for types where type invariants are enforced.
For `NonZeroU32`, this means we never return a `0` value.
The assertion we wrote in this harness was just an extra check we added to demonstrate this fact.

## Other kinds of nondeterministic variables

There's nothing special about `kani::any()` except that Kani ships with it only implemented for types where we can guarantee safety.

> Notably, however, there is no implementation for `Vec<T>`, for example.

Reviewer: But we just added `any_vec`.

The trouble with a nondeterministic vector is that you usually need to _bound_ the size of the vector, as we saw in the last chapter.
There are no arguments provided to `kani::any()` to indicate an upper bound.

Likewise, `kani::any()` is implemented for `[T; N]` (if implemeneted for `T`).

> But, it is _not_ implemented for `&[T]`.

Reviewer: This doesn't make a lot of sense to me. This is a fat pointer. Maybe we should say that we can't handle non-deterministic pointers or that non-deterministic objects that involve memory allocation require special handling?

Again, there's no upper bound information we can make use of.

This does not mean you cannot have a nondeterministic vector.

> You just have to construct it.

Reviewer: But how? I think this is missing. We could document the Arbitrary type here. That is a very safe trait.
